c/tskit/tables.c

```diff
@@ -8105,6 +8105,285 @@ tsk_table_collection_subset(
+        goto out;
+    }
+    ret = tsk_table_collection_init(&other_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    i = 0;
+    for (k = 0; k < (tsk_id_t) other_num_nodes; k++) {
+        if (node_mapping[k] != TSK_NULL) {
+            self_nodes[i] = node_mapping[k];
+            other_nodes[i] = k;
+            i++;
+        }
+    }
+
+    // TODO: strict sort before checking equality
+    ret = tsk_table_collection_copy(self, &self_copy, TSK_NO_INIT);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_copy(other, &other_copy, TSK_NO_INIT);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&other_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&self_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&self_copy, self_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&other_copy, other_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = !tsk_table_collection_equals(&self_copy, &other_copy);
+    if (ret != 0) {
+        ret = TSK_ERR_DIFF_HISTORIES;
+        goto out;
+    }
+
+out:
+    tsk_table_collection_free(&self_copy);
+    tsk_table_collection_free(&other_copy);
+    tsk_safe_free(other_nodes);
+    tsk_safe_free(self_nodes);
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_union(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *node_mapping, tsk_size_t num_nodes, tsk_flags_t options)
+{
+    int ret = 0;
+    tsk_id_t k, i, new_ind, new_pop, new_parent, new_child;
+    tsk_size_t num_shared_nodes = 0;
+    tsk_node_t node;
+    tsk_individual_t ind;
+    tsk_population_t pop;
+    tsk_edge_t edge;
+    tsk_mutation_t mut;
+    tsk_site_t site;
+    tsk_id_t *node_map = NULL;
+    tsk_id_t *individual_map = NULL;
+    tsk_id_t *population_map = NULL;
+    tsk_id_t *site_map = NULL;
+    bool add_populations = !(options & TSK_UNION_NO_ADD_POP);
+    bool check_shared_portion = !(options & TSK_UNION_NO_CHECK_SHARED);
+
+    /* node_mapping relates id in other (index) to equivalent node in self or
+     * TSK_NULL if not present */
+    if (num_nodes != (tsk_size_t) other->nodes.num_rows) {
+        ret = TSK_ERR_LEN_MAP;
+        goto out;
+    }
+
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        if (node_mapping[k] >= (tsk_id_t) self->nodes.num_rows) {
+            ret = TSK_ERR_BAD_MAP;
+            goto out;
+        }
+        if (node_mapping[k] != TSK_NULL)
+            num_shared_nodes++;
+    }
+
+    // checking shared equality
+    if (check_shared_portion) {
+        ret = tsk_check_subset_equality(
+            self, other, node_mapping, num_nodes, num_shared_nodes);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+    // maps relates id in other (index) to new id in self
```

Comment:
I'm not sure what the (index) refers to here? Also, I think it helps to write comments grammatically, using punctuation etc. It's all about communication. I'd say
```
// Maps relating the IDs in other to the new IDs in self.
```